iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+
+/// <summary>
+/// Update a batch of properties.
+/// 
+/// Properties under the root component:
+/// "reported": {
+///     "temperature": 21.3,
+///     "humudity": 60
+/// }
+/// 
+/// Properties under a specified component:
+/// "reported": {
+///     "thermostat1": {
+///         "__t": "c",
+///         "temperature": 21.3,
+///         "humudity": 60
+///     }
+/// }
+/// </summary>
+/// <param name="properties">The twin properties and values to update.</param>
+/// <param name="componentName">The name of the component in which the properties are defined. Can be null for properties defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertiesAsync(IDictionary<string, dynamic> properties, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Respond to a writable property request received from the service.
+/// 
+/// Property under the root component:
+/// "reported": {
+///     "targetTemperature": {
+///         "value": 21.3,
+///         "ac": 200,
+///         "av": 3,
+///         "ad": "complete"
+///     }
+/// }
+/// 
+/// Property under a specified component:
+/// "reported": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": {
+///       "value": 23,
+///       "ac": 200,
+///       "av": 3,
+///       "ad": "complete"
+///     }
+///   }
+/// }
+/// </summary>
+/// <param name="propertyName">The name of the twin property to update.</param>
+/// <param name="propertyValue">The value of the twin property to update.</param>
+/// <param name="componentName">The name of the component in which the property is defined. Can be null for property defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, WritableProperty propertyValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Subscribe to writable property event for all properties under any component.
+///
+/// Payload received for the root component:
+/// "desired" :
+/// {
+///   "targetTemperature" : 21.3,
+///   "targetHumidity" : 80
+/// },
+/// "$version" : 3
+///
+/// Payload received for a specific component:
+/// "desired": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": 21.3,
+///     "targetHumidity": 80
+///   }
+/// },
+/// "$version" : 3
+/// 
+/// </summary>
+/// <param name="propertyActionAsTwinCollection">The action to be taken when a writeable property event is received.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public void SubscribeToWritablePropertyEvent(Action<TwinCollection> propertyActionAsTwinCollection, CancellationToken cancellationToken = default);
```

Comment:
We can add another API for managing per-component property updates (based on customer feedback)

Reply:
Yes I think we should have one  with componentName as parameter

Reply:
I agree with this. If we ship with the ability per component name we will need to create a dispatcher of sorts.